pandas/core/internals.py

```diff
@@ -3748,18 +3753,44 @@ def comp(s):
             for i, (s, d) in enumerate(zip(src_list, dest_list)):
                 new_rb = []
                 for b in rb:
-                    if b.dtype == np.object_:
-                        convert = i == src_len
-                        result = b.replace(s, d, inplace=inplace, regex=regex,
-                                           mgr=mgr, convert=convert)
-                        new_rb = _extend_blocks(result, new_rb)
+                    # regular expression support needs to be improved.
+                    # If the replacement for the previous pattern
+                    # matches the next pattern, the value will be replaced
+                    # again with a different value from dest_list
+                    # i.e. when values is ['a', 'b']
+                    # src_list: [r'a*', r'b*'], dest_list: ['b', 'a']
+                    # result will be ['b', b'] after searching for pattern r'a'
+                    # and then changed to ['a', 'a'] for pattern r'b*'
+                    if regex:
+                        if b.dtype == np.object_:
+                            convert = i == src_len
+                            result = b.replace(s, d, inplace=inplace,
+                                               regex=regex,
+                                               mgr=mgr, convert=convert)
+                            new_rb = _extend_blocks(result, new_rb)
```

Comment:
instead of this, can you add a new (private) method on the Block itself (and then override for object dtype). It will be much cleaner code and this part becomes really generic.